Step17 optionals

ShihaoXu0420 opened the pull request:
## Requirements

- Prevent other user to get access to user's tasks (authentication)
- Only allow unique username created

## Behaviors

- [x] Prevent show other user's task
- [x] Prevent show other user's task's edit page
- [x] Prevent put the form to other user's task endpoint
- [x] Prevent access to login/signup page when logged in 
  => All redirected to index page
  <video src="https://github.com/Fablic/training/assets/167514427/f3b1d827-58db-4dcb-a665-2c51483272d8" width=400 />

---

- [x] Prevent duplicated username created
   <video src="https://github.com/Fablic/training/assets/167514427/33cec01b-e085-4d09-8c84-4e93c50b15bb" width=400 />

ShihaoXu0420 committed:
Handle invalid access to other user's tasks

ShihaoXu0420 committed:
Make username unique

ShihaoXu0420 committed:
Handle access to login/signup page when logged in

ShihaoXu0420 committed:
Fix log out bug

rr-mdalauddin-alamin commented on myapp/app/controllers/sessions_controller.rb:
```diff
@@ -1,5 +1,6 @@
 class SessionsController < ApplicationController
   def new
+    redirect_to tasks_path, notice: 'Already logged in!' if logged_in?
```
instead of checking in the method, how about creating a helper method and using in the `before_action`. 
or you can create a middlewhere and always redirect to next clouser if true.

ShihaoXu0420 replied:
Fixed it in eb925c6

rr-mdalauddin-alamin commented on myapp/app/models/user.rb:
```diff
@@ -5,4 +5,5 @@ class User < ApplicationRecord
 
   has_many :tasks, dependent: :destroy
   validates :username, format: { with: /\A[^\s]*\z/, message: 'should not contain space!' }
+  validates :username, uniqueness: { message: 'duplicated. Please use another username!' }
```
test cases would be saver based on validations

ShihaoXu0420 replied:
Added test case in 0b32528

rr-mdalauddin-alamin commented on myapp/app/controllers/users_controller.rb:
```diff
@@ -1,5 +1,6 @@
 class UsersController < ApplicationController
   def new
+    redirect_to tasks_path, notice: 'Already logged in!' if logged_in?
```
same as https://github.com/Fablic/training/pull/1667/files#r1632505976

ShihaoXu0420 replied:
Fixed it in eb925c6

ShihaoXu0420 committed:
Merge branch 'ShihaoXu0420' into ShihaoXu0420-step17-optionals

ShihaoXu0420 committed:
Add before_action ti sessions controller

ShihaoXu0420 committed:
Added test case for username uniqueness

ksak19 commented on myapp/app/controllers/sessions_controller.rb:
It's not related to changes this time but...
For `#signup_new` and `#signup_create`, I recommend to move it to `UsersController` and it should be `#new` and `#create`
https://github.com/Fablic/training/blob/294cbf53a198363877100262419da8f020a3456f/myapp/app/controllers/users_controller.rb

Please refer to this way of thinking as well 😄 
https://postd.cc/how-dhh-organizes-his-rails-controllers/
It's not the absolute answer but I like this idea

ShihaoXu0420 replied:
Actually I was putting them in `UsersController`...
https://github.com/Fablic/training/pull/1666#discussion_r1632531816

ksak19 replied:
hmm, I have different feeling and your previous implementation is good, imo. `SessionsController` is responsible for Login/Logout and `UsersController` can be responsible for creating user, meaning Signup. Login/Logout and Signup is totally different things, I would say

ShihaoXu0420 replied:
Yeah and I remember I was referring to [this tutorial](https://railstutorial.jp/chapters/log_in_log_out?version=4.2#sec-login_form) when developing which also takes the same implementation

ShihaoXu0420 replied:
So should I revert it back to origin one?

ksak19 replied:
Up to you! But if you want to do that, you should create another PR since it's different theme from this PR 😄 
If you will do it, I can review the PR again so feel free to ask me

ksak19 replied:
so maybe it's good to keep it as it is for now and fix them later

ShihaoXu0420 replied:
Ok let me fix it later then

ksak19 commented on myapp/app/controllers/tasks_controller.rb:
```diff
@@ -73,4 +75,14 @@ def task_params
+
+  def authorize_task
+    if find_tasks_by_user.find_by(id: params[:id]).nil?
+      redirect_to tasks_path, notice: 'No task found!'
```
📝 redirect is fine but it can be 403 error ideally

ShihaoXu0420 replied:
I understand! But pls just let me stick to the original one.

ksak19 reviewed (commented):
他人のタスクを開こうとしたときに意図した通りredirectされるテストもあると良いなと思いました！
`spec/system/tasks_spec.rb` かな

その他2点コメントしましたが、いま修正しなくてもいいと思ってます！

rr-mdalauddin-alamin commented:
@ShihaoXu0420 following this [suggestion](https://github.com/Fablic/training/pull/1667#pullrequestreview-2112061294) you are going to write test for this right?
I'll review again, once you finish then?

ShihaoXu0420 commented:
> @ShihaoXu0420 following this [suggestion](https://github.com/Fablic/training/pull/1667#pullrequestreview-2112061294) you are going to write test for this right? I'll review again, once you finish then?

Yes! Once I finish writing the test!

ksak19 commented:
@ShihaoXu0420 
I can review

ShihaoXu0420 committed:
Add tests for redirection for invalid access

ShihaoXu0420 commented:
@ksak19 
Added two test for invalid request to task detail page and task edit page
![image](https://github.com/Fablic/training/assets/167514427/e10630df-854a-43d4-b703-e9ffd53a84fd)

![image](https://github.com/Fablic/training/assets/167514427/3aafe165-4e35-44a4-8ec2-c03d23d719da)

* I only tested the cases which I added this time, other cases will fail due to the changes by previous steps (i18n, adding user...) but fixing all of them will take too much time and I think its more important to move on to next step. So I will just keep them.

ksak19 reviewed (approved):
![LGTM](https://image.lgtmoon.dev/251061)
